### PR TITLE
Implement instanced WebGL drawing calls

### PR DESCRIPTION
**Pull request by nox**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/20364)

### CODE REVIEW

File: components/script/dom/webglrenderingcontext.rs [hunk 1]
```diff
@@ -2378,20 +2479,30 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
...
-            return jsval.get();
+            return result.get();
+        }
+
+        if !self.extension_manager.is_get_vertex_attrib_name_enabled(pname) {
+            self.webgl_error(WebGLError::InvalidEnum);
+            return NullValue();
+        }
+
+        if index == 0 && pname == ANGLEInstancedArraysConstants::VERTEX_ATTRIB_ARRAY_DIVISOR_ANGLE {
```
Comment (emilio) [inline components/script/dom/webglrenderingcontext.rs]: This needs a comment, pointing or quoting the relevant specification paragraph.

File: components/script/dom/webglrenderingcontext.rs [hunk 2]
```diff
@@ -1160,6 +1163,104 @@ impl WebGLRenderingContext {
...
+        match mode {
+            constants::POINTS | constants::LINE_STRIP |
+            constants::LINE_LOOP | constants::LINES |
+            constants::TRIANGLE_STRIP | constants::TRIANGLE_FAN |
+            constants::TRIANGLES => {},
+            _ => {
+                return self.webgl_error(InvalidEnum);
+            }
+        }
+        if count < 0 || offset < 0 {
```
Comment (emilio) [inline components/script/dom/webglrenderingcontext.rs]: This is missing the validation for `primcount`:

>  GL_INVALID_VALUE is generated if count or primcount are negative.

Comment (emilio) [reply components/script/dom/webglrenderingcontext.rs]: (In general it'd be nice if there were spec quotes around, wdyt?)

File: components/script/dom/webglrenderingcontext.rs [hunk 3]
```diff
@@ -1160,6 +1163,104 @@ impl WebGLRenderingContext {
...
+            return self.webgl_error(InvalidValue);
+        }
+        let type_size = match type_ {
+            constants::UNSIGNED_BYTE => 1,
+            constants::UNSIGNED_SHORT => 2,
```
Comment (emilio) [inline components/script/dom/webglrenderingcontext.rs]: https://www.khronos.org/registry/OpenGL-Refpages/es3.0/html/glDrawElementsInstanced.xhtml also says to allow `UNSIGNED_INT`, is that not available in WebGL, or just an oversight? In any case comment deserved.

File: components/script/dom/webglrenderingcontext.rs [hunk 4]
```diff
@@ -1160,6 +1163,104 @@ impl WebGLRenderingContext {
...
+            _ => return self.webgl_error(InvalidEnum),
+        };
+        if offset % type_size != 0 {
+            return self.webgl_error(InvalidOperation);
+        }
```
Comment (emilio) [inline components/script/dom/webglrenderingcontext.rs]: This is missing: https://www.khronos.org/registry/webgl/specs/latest/2.0/#ATTRIBUTE_TYPE_MATCH.

Though it's kinda hard I think, we need to retain more attribute information for it. Maybe file and leave a TODO?

**Review by emilio — COMMENTED** [review]
Also, tests? :)

**Comment by nox** [comment]
Closing this for now because I want to fix the issues with DrawArrays first.

**Pull request closed without merging**